pco/src/histograms.rs

```diff
@@ -44,29 +44,22 @@ pub struct HistogramBin<L: Latent> {
 }
 
 fn slice_min<L: Latent>(latents: &[L]) -> L {
-  let mut min0 = L::MAX;
-  let mut min1 = L::MAX;
-  for i in (0..latents.len()).skip(1).step_by(2) {
-    min0 = min(min0, latents[i - 1]);
-    min1 = min(min1, latents[i]);
-  }
-  if latents.len() % 2 == 1 {
-    min0 = min(min0, latents.last().cloned().unwrap());
-  }
-  min(min0, min1)
+  latents
```

Comment:
I find that we can do `latents.iter().cloned().min().unwrap()` and still get SIMD. I think the difference between this and `*latents.iter().min().unwrap()` is that the latter operates on references to latents instead of values.

Reply:
Didn't realise a had to clone before aggregating for it to vectorize. Though now that I look at it, it's what I was doing with `fold`.

One of the reasons I was using fold is that it was functionally equivalent to the previous implementation in the way it would handle an empty array. I also kinda like that it avoids `unwrap`. `fold` also appears to produce slightly fewer instructions, though I doubt it makes any measurable difference.

If we make `slice_min` and `slice_max` panic on empty array, then I think `slice_min_max` should panic too. If we do it like this
``` rust
fn slice_min<L: Latent>(latents: &[L]) -> L {
  latents.iter().cloned().min().unwrap()
}

fn slice_max<L: Latent>(latents: &[L]) -> L {
  latents.iter().cloned().max().unwrap()
}

fn slice_min_max<L: Latent>(latents: &[L]) -> (L, L) {
  latents.iter().cloned().fold(
    (L::MAX, L::ZERO),
    |(min_val, max_val), val| (min(min_val, val), max(max_val, val)),
  )
}
```
then the three functions handle empty arrays differently, which isn't ideal imho. I don't know if there's a clever way to do `slice_min_max` that makes it function like the two others.

Reply:
I don't think you should worry about it, since these functions are private to this module and only used on nonempty slices. But if this still bothers you, you could do e.g. `.unwrap_or(L::MAX)` to make slice_min behave like `slice_min_max`.

Reply:
Personally, I think the current implementation where all three functions use `fold` is cleaner than the alternatives, as it sticks to one way of doing things.

If we got rid of the functions and aggregated inline I think using the built-in `min` and `max` funcitons would be cleaner. However, as long as we use the custom `slice_` functions, which I think makes the code more readable overall, I like that they're consistent, both functionally and implementation wise.

It's your call of course :-)
If you prefer something else, feel free to push the changes.

Reply:
Ok, seems reasonable. I just made a tiny simplification.